### PR TITLE
Weak reference

### PR DESCRIPTION
**Pull request by adam-fowler**
I've added this to avoid cyclical references.

This is an implementation of [`WeakRef`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakRef) which doesn't seem to be available yet.
It stores references in a dictionary, to avoid cyclical references. 
I have used it in one place so far `PackageWatcher`
I also fixed the Workspace cyclical reference, by removing `ExtensionContext` from it. We don't use it anywhere at the moment.

### CODE REVIEW

File: src/FolderContext.ts [hunk 1]
```diff
@@ -29,7 +30,6 @@ export class FolderContext implements vscode.Disposable {
     ) {
         if (this.isRootFolder) {
             this.packageWatcher = new PackageWatcher(this, workspaceContext);
-            this.packageWatcher.install();
```
Comment (0xTim) [inline src/FolderContext.ts:32]: What does this have to do with weak references?

Comment (adam-fowler) [reply src/FolderContext.ts:32]: The install and constructor have been combined in this PR. I guess it isn't completely related

**Review by 0xTim — APPROVED** [review]

**Comment by adam-fowler** [comment]
Most modern javascript engines garbage collection can deal with cyclical references so this shouldn't be an issue.
